packages/starlight/utils/virtual-page.ts

```diff
@@ -0,0 +1,169 @@
+			frontmatter: StarlightVirtualFrontmatter;
+		}
+>;
+
+/**
+ * A docs entry used for virtual pages meant to be rendered by plugins and which is safe to cast
+ * to a `StarlightDocsEntry`.
+ * A virtual docs entry cannot be rendered like a content collection entry.
+ */
+type VirtualDocsEntry = Omit<StarlightDocsEntry, 'id' | 'render'> & {
+	/**
+	 * The unique ID for this virtual page which cannot be inferred from codegen like content
+	 * collection entries.
+	 */
+	id: string;
+};
+
+export function generateVirtualRouteData({
+	props,
+	url,
+}: {
+	props: VirtualPageProps;
+	url: URL;
+}): StarlightRouteData {
+	const { isFallback, frontmatter, slug, ...routeProps } = props;
+	const virtualFrontmatter = getVirtualFrontmatter(frontmatter);
+	const id = `${stripLeadingAndTrailingSlashes(slug)}.md`;
+	const localeData = slugToLocaleData(slug);
+	const sidebar = props.sidebar ?? getSidebar(url.pathname, localeData.locale);
+	const headings = props.headings ?? [];
+	const virtualEntry: VirtualDocsEntry = {
+		id,
+		slug,
+		body: '',
+		collection: 'docs',
+		data: {
+			...virtualFrontmatter,
+			sidebar: {
+				attrs: {},
+				hidden: false,
+			},
+		},
+	};
+	const entry = virtualEntry as StarlightDocsEntry;
+	const entryMeta: StarlightRouteData['entryMeta'] = {
+		dir: props.dir ?? localeData.dir,
+		lang: props.lang ?? localeData.lang,
+		locale: localeData.locale,
+	};
+	const editUrl = virtualFrontmatter.editUrl ? new URL(virtualFrontmatter.editUrl) : undefined;
+	const lastUpdated =
+		virtualFrontmatter.lastUpdated instanceof Date ? virtualFrontmatter.lastUpdated : undefined;
+	const routeData: StarlightRouteData = {
+		...routeProps,
+		...localeData,
+		id,
+		editUrl,
+		entry,
+		entryMeta,
+		hasSidebar: props.hasSidebar ?? entry.data.template !== 'splash',
+		headings,
+		labels: useTranslations(localeData.locale).all(),
+		lastUpdated,
+		pagination: getPrevNextLinks(sidebar, config.pagination, entry.data),
+		sidebar,
+		slug,
+		toc: getToC({
+			...routeProps,
+			...localeData,
+			entry,
+			entryMeta,
+			headings,
+			id,
+			locale: localeData.locale,
+			slug,
+		}),
+	};
+	if (isFallback) {
+		routeData.isFallback = true;
+	}
+	return routeData;
+}
+
+/** Validates the virtual frontmatter properties from the props received by a virtual page. */
+function getVirtualFrontmatter(frontmatter: StarlightVirtualFrontmatter) {
+	// This needs to be in sync with ImageMetadata.
+	// https://github.com/withastro/astro/blob/cf993bc263b58502096f00d383266cd179f331af/packages/astro/src/assets/types.ts#L32
+	return StarlightVirtualFrontmatterSchema({
+		image: () =>
+			z.object({
+				src: z.string(),
+				width: z.number(),
+				height: z.number(),
+				format: z.union([
+					z.literal('png'),
+					z.literal('jpg'),
+					z.literal('jpeg'),
+					z.literal('tiff'),
+					z.literal('webp'),
+					z.literal('gif'),
+					z.literal('svg'),
+					z.literal('avif'),
+				]),
+			}),
+	}).parse(frontmatter);
```

Comment:
One thing I’m thinking about here is what happens to a user’s schema extensions in `src/content/config.ts`. We don’t really have access to them and this `parse()` call will only use the default schema (although the type will be inferred from their extensions if I followed everything correctly, because it’s based on `StarlightDocsEntry` and therefore `CollectionEntry`).

There are some quick solutions that get us _some_ of the way there, like:

```suggestion
	})
		.passthrough()
		.parse(frontmatter);
```

That would at least preserve additional values passed to the frontmatter prop rather than stripping them. But it wouldn’t solve cases where the frontmatter schema extension includes default values or transforms.

Not sure how big of a concern this is — we could add it as a future to-do I guess. In theory, importing `collections` from a user’s `src/content/config.ts` is do-able and `collections.docs.schema` could give us the full schema instead of using `StarlightFrontmatterSchema` directly.

Reply:
Damn, total oversight on my end, thanks for pointing that out.

The `passthrough()` is indeed a solution and you are correct, the user, e.g. in an override would get the value properly typed (just tested it). Altho, now this makes me think of the same issue from a plugin author perspective:

- A plugin want to use virtual routes.
- This plugin also provide a schema that the user should extend from which add a mew optional value.
- When using a virtual route, the plugin author will fill the props based on `VirtualPageProps` and for the `frontmatter` property, it's based on `StarlightVirtualFrontmatter` which is inferred from the `StarlightVirtualFrontmatterSchema` which extends the `StarlightFrontmatterSchema` which has no knowledge about any potential extension.

If the plugin author want to provide this new additional value, they would get an error when doing something like this:

```ts
const props = {
  slug: 'custom-page/example',
  frontmatter: {
    title: 'My custom page',
    customPluginValue: 'foo',
    // ^^^^^^^^^^^^^^^^^^^^^^
  },
} satisfies VirtualPageProps;
```

I'm wondering if `VirtualPageProps` should be a generic type that would allow the plugin author to provide their own extra props for the frontmatter. Something like:

```ts
type PluginExtend = z.infer<typeof PluginExtendSchema>;

const props = {
  slug: 'custom-page/example',
  frontmatter: {
    title: 'My custom page',
    customPluginValue: 'foo',
  },
} satisfies VirtualPageProps<PluginExtend>;
```

And we would just need to tweak the `VirtualPageProps` type to allow for this generic type, e.g. `VirtualPageProps<T = {}>` and the frontmatter key would become `frontmatter: StarlightVirtualFrontmatter & T;`.

Any thoughts?

Reply:
Oh right, just seeing that my assumption about types was wrong above, because `StarlightVirtualFrontmatter` is inferred from the schema, not `StarlightDocsEntry` as I thought (which also makes sense because `StarlightDocsEntry` is an output type not the input type).

I think a generic here might be kind of annoying to use. I think we can probably solve this somehow, perhaps with a virtual module that re-exports a user’s content collections config (or a default config if that is not found):

```ts
// virtual:starlight/collection-config
import { defineCollection } from 'astro:content';
import { docsSchema, i18nSchema } from '@astrojs/starlight/schema';

let userCollections;
try {
	userCollections = await import('resolved/path/to/src/content/config.ts').collections;
} catch {}
if (!userCollections) {
	userCollections = {
		docs: defineCollection({ schema: docsSchema() }),
		i18n: defineCollection({ type: 'data', schema: i18nSchema() }),
	};
}
export const collections = userCollections;
```

I’ll also ask if exporting the [`ContentConfig` type](https://github.com/withastro/astro/blob/a3df9d83ca92abb5f08f576631019c1604204bd9/packages/astro/content-types.template.d.ts#L186) from `astro:content` is possible — that would make it easier to get that type in our code. Currently it generates a line like this in `.astro/types.d.ts`:

```ts
type ContentConfig = typeof import("../src/content/config");
```

If that was exported, we could easily type our virtual module as something like:

```ts
declare module 'virtual:starlight/collection-config' {
	export const collections: import('astro:content').ContentConfig['collections'];
}
```

Reply:
Thanks for the input, always appreciated. Before jumping into the code and to make sure I get it right:

- We would import the schema from the virtual module and use `collections.docs.schema` to actually define the `StarlightVirtualFrontmatterSchema` and omit/extend it like we actually do with the `StarlightFrontmatterSchema`.
- We would then have the user schema extension when using `StarlightVirtualFrontmatterSchema` to do the parsing.
- A user using the `VirtualPageProps` type in its own project would have proper typing for the virtual frontmatter including the schema extension.

Altho, I don't think this addresses the issue I mentioned above of a plugin author developing a plugin in isolation  and getting error when trying to add custom property to the frontmatter, correct? We would need something else for that (and I have no idea yet what that could be if not a generic)

Reply:
Yes, that all sounds right to me!

> I don't think this addresses the issue I mentioned above of a plugin author developing a plugin in isolation and getting error when trying to add custom property to the frontmatter, correct?

Yes, I think this a similar issue to Starlight’s where types based on `astro:content` are not available (except for the fact we kind of hackily have them around from the docs site thanks to some specifics of our monorepo set-up). That said, I _think_ because the type in question here would fail to resolve, it would be `any` for plugin authors, so shouldn’t cause errors?

Reply:
Just played with it a bit locally and here are my first impressions:

- Getting hold of the schema is relatively easy indeed.

- We cannot use it yet to validate the data as when extended, we are using an [intersection](https://github.com/withastro/starlight/blob/main/packages/starlight/schema.ts#L156) which prevents us to refine the schema as stated in the Zod docs:

  > `A.and(B)` returns a less useful `ZodIntersection` instance that lacks common object methods like `pick` and `omit`.

  Not sure yet if we can work around this, I'll need to dig a bit more.

- > That said, I think because the type in question here would fail to resolve, it would be `any` for plugin authors, so shouldn’t cause errors?

  I think this is correct but I wonder if it's better or worse. As a plugin author, I would at least expect to get auto completion/errors for at least the known properties. :thinking: Need to think about this a bit more.

Reply:
Small update after more investigations, it looks like the following could work:

```ts
const StarlightVirtualFrontmatterSchema = (context: SchemaContext) => {
  const docsSchema = collections.docs.schema!;
  const schema =
    typeof docsSchema === "function" ? docsSchema(context) : docsSchema;

  return schema
    .and(
      z.object({
        editUrl: z.union([z.string().url(), z.literal(false)]).default(false),
      })
    )
    .transform((frontmatter) => {
      const { sidebar, ...others } = frontmatter;
      return others;
    });
};
```

After applying your change from the Astro PR and adding the types for `virtual:starlight/collection-config`, this works as expected **in the Starlight code base**. I get proper typing when the Docs schema is not extended and the same when extended, I can see the extended properties and everything. Same goes for the validation, it works as expected from my initial tests.

So now, even if I simplify the schema by returning it unmodified with no `.and()` and `.transform()` and directly exporting its type:

```ts
export type StarlightVirtualFrontmatter = z.input<
  ReturnType<typeof StarlightVirtualFrontmatterSchema>
>;
```

And re-exporting it from `packages/starlight/props.ts` (or in the virtual page component like it's currently done in the PR), I cannot get the type to resolve anywhere even in the Starlight docs when importing it. It's always `any`.

I'll keep investigating and see if there are something that could help me figure out why this is not resolving properly, maybe I'm missing something obvious, not sure.

If you have any idea in the meantime, I'm all ears and I'll keep you posted if I find anything.

Reply:
I’m not sure I follow all the ins and outs of code at this point, but if you want to push some WIP code for me to look at, I’d be happy to try playing with it locally!

Reply:
Sure, I just pushed a [WIP commit](https://github.com/withastro/starlight/pull/1175/commits/2f6905ed53472f8741cca82321d3b794a36c2856) (tests are breaking, did not look into it yet, just trying to get the basics working).

- The main issue is in `docs/src/components/about-astro.astro` where the `StarlightVirtualFrontmatter` imported type is `any`.
- In `packages/starlight/utils/virtual-page.ts` it is not tho.

Note that it needs the change from https://github.com/withastro/astro/pull/9840

Reply:
> It looks like this starts happening when the file being imported by the virtual module in build mode also import astro:content.

You mean the `import { defineCollection } from 'astro:content';` in the `virtual:starlight/collection-config` module? We could probably get rid of that if so — `defineCollection()` is an identity function just used for type checking, I think, so `defineCollection({ schema: foo })` is actually identical to `{ schema: foo }`.

(Maybe I misunderstood though — I’m too sleepy on this train and also agree you deserve a break 😁)

Reply:
I think it's more the `import { defineCollection } from 'astro:content';` in the `src/content/config.ts` file which is imported by the virtual module. I can successfully import another file from the virtual module, e.g. `src/content/test.ts` which contains random code, but as soon as this file import `astro:content` like the config is doing, the build will not build any page (but not fail).

I'll investigate more tomorrow. Have a good ride home!

Reply:
Ok, finally fixed and we should have virtual pages in the preview. For one part, I definitely needed a break, for the other part, I'm not quite sure yet but I had to move outside of the virtual module the fallback to our default schema if we could not import the user config.

I pushed the changes and also addressed your other feedback regarding type naming, I don't care at all, just an old habit ^^